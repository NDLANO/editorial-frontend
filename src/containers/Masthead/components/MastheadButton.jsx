/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Cross } from '@ndla/icons/action';
import { Menu } from '@ndla/icons/common';
import { injectT } from '@ndla/i18n';
import { css } from 'react-emotion';
import { colors, fonts, spacing, misc } from '@ndla/core';

const buttonStyle = css`
  background: transparent;
  padding: ${spacing.small} ${spacing.normal};
  ${fonts.sizes(16, 1.625)};
  font-weight: ${fonts.weight.normal};
  border-radius: ${misc.borderRadius};
  border: 2px solid ${colors.brand.primary};
  color: ${colors.brand.primary};
  transition: 200ms all ease;
  display: flex;
  align-items: center;

  svg {
    margin-right: ${spacing.xsmall};
  }
  &:hover {
    border: 2px solid transparent;
    background: ${colors.brand.primary};
    color: ${colors.white};
  }
  &:active,
  &:focus {
    border: 2px solid ${colors.brand.lighter};
    background: ${colors.white};
    color: ${colors.brand.primary};
  }
`;

const crossCss = css`
  width: 22px;
  height: 22px;
`;

const MastheadButton = ({ children, color, minWidth, open, onClick, t }) => (
  <button type="button" onClick={onClick} css={buttonStyle}>
    {open ? (
      <>
        <Cross css={crossCss} />
        <span>{t('masthead.closeMenu')}</span>
      </>
    ) : (
      <>
        <Menu />
        <span>{t('masthead.menu')}</span>
      </>
    )}
  </button>
);

MastheadButton.propTypes = {
  color: PropTypes.string,
  minWidth: PropTypes.number,
  open: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
};

MastheadButton.defaultProps = {
  color: colors.brand.grey,
  minWidth: 0,
};

export default injectT(MastheadButton);
