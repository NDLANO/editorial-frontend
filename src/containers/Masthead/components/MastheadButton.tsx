/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import { Cross } from '@ndla/icons/action';
import { Menu } from '@ndla/icons/common';
import { useTranslation } from 'react-i18next';
import { colors, fonts, spacing, misc } from '@ndla/core';
import { css } from '@emotion/core';

const buttonStyle = css`
  background: transparent;
  padding: ${spacing.small} ${spacing.normal};
  ${fonts.sizes(16, 1.625)};
  font-weight: ${fonts.weight.normal};
  border-radius: ${misc.borderRadius};
  border: 2px solid ${colors.brand.primary};
  color: ${colors.brand.primary};
  transition: 200ms all ease;
  display: flex;
  align-items: center;

  svg {
    margin-right: ${spacing.xsmall};
  }
  &:hover {
    border: 2px solid transparent;
    background: ${colors.brand.primary};
    color: ${colors.white};
  }
  &:active,
  &:focus {
    border: 2px solid ${colors.brand.lighter};
    background: ${colors.white};
    color: ${colors.brand.primary};
  }
`;

const crossCss = css`
  width: 22px;
  height: 22px;
`;

interface Props {
  open: boolean;
  onClick: () => void;
}

const MastheadButton = ({ open, onClick }: Props) => {
  const { t } = useTranslation();
  return (
    <button type="button" onClick={onClick} css={buttonStyle}>
      {open ? (
        <>
          <Cross css={crossCss} />
          <span>{t('masthead.menu.close')}</span>
        </>
      ) : (
        <>
          <Menu />
          <span>{t('menu.title')}</span>
        </>
      )}
    </button>
  );
};

export default MastheadButton;
