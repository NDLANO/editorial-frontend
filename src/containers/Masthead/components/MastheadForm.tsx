/**
 * Copyright (c) 2024-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ComponentPropsWithRef, FormEvent, forwardRef } from "react";
import { useTranslation } from "react-i18next";
import { useNavigate } from "react-router-dom";
import { styled } from "@ndla/styled-system/jsx";
import { isValidLocale } from "../../../i18n";
import { fetchNewArticleId } from "../../../modules/draft/draftApi";
import { fetchNode } from "../../../modules/nodes/nodeApi";
import { resolveUrls } from "../../../modules/taxonomy/taxonomyApi";
import { isNDLAFrontendUrl } from "../../../util/htmlHelpers";
import { routes } from "../../../util/routeHelpers";
import { useTaxonomyVersion } from "../../StructureVersion/TaxonomyVersionProvider";

interface Props extends ComponentPropsWithRef<"form"> {
  query: string;
  onSubmit: () => void;
}

const StyledForm = styled("form", {
  base: {
    width: "100%",
  },
});

export const MastheadForm = forwardRef<HTMLFormElement, Props>(({ children, query, onSubmit, ...props }, ref) => {
  const { i18n } = useTranslation();
  const { taxonomyVersion } = useTaxonomyVersion();
  const navigate = useNavigate();
  const handleNodeId = async (nodeId: number) => {
    try {
      const newArticle = await fetchNewArticleId(nodeId);
      navigate(routes.editArticle(newArticle.id, "standard"));
    } catch (error) {
      navigate(routes.notFound);
    }
  };

  const handleTaxonomyId = async (taxId: string) => {
    try {
      const taxElement = await fetchNode({ id: taxId, taxonomyVersion });
      const arr = taxElement.contentUri?.split(":");
      if (arr) {
        const id = arr[arr.length - 1];
        navigate(routes.editArticle(parseInt(id), "standard"));
      }
    } catch (error) {
      navigate(routes.notFound);
    }
  };

  const handleUrlPaste = (frontendUrl: string) => {
    // Removes search queries before split
    const ndlaUrl = frontendUrl.split(/\?/)[0];
    // Strip / from end if topic
    const cleanUrl = ndlaUrl.endsWith("/")
      ? ndlaUrl.replace("/subjects", "").slice(0, -1)
      : ndlaUrl.replace("/subjects", "");
    const splittedNdlaUrl = cleanUrl.split("/");

    const urlId = splittedNdlaUrl[splittedNdlaUrl.length - 1];

    if (
      !urlId.includes("urn:topic") &&
      Number.isNaN(parseFloat(urlId)) &&
      !splittedNdlaUrl.find((e) => e.match(/subject:*/)) === undefined
    ) {
      return;
    }
    if (urlId.includes("urn:topic")) {
      handleTopicUrl(urlId);
    } else if (splittedNdlaUrl.includes("node")) {
      handleNodeId(parseInt(urlId));
    } else if (splittedNdlaUrl.find((e) => e.match(/subject:*/))) {
      handleFrontendUrl(cleanUrl);
    } else {
      navigate(routes.editArticle(parseInt(urlId), "standard"));
    }
  };

  const handleTopicUrl = async (urlId: string) => {
    try {
      const topicArticle = await fetchNode({
        id: urlId,
        language: i18n.language,
        taxonomyVersion,
      });
      const arr = topicArticle.contentUri?.split(":") ?? [];
      const id = arr[arr.length - 1];
      navigate(routes.editArticle(parseInt(id), "topic-article"));
    } catch {
      navigate(routes.notFound);
    }
  };

  const handleFrontendUrl = async (url: string) => {
    const { pathname } = new URL(url);
    const paths = pathname.split("/");
    const path = isValidLocale(paths[1]) ? paths.slice(2).join("/") : pathname;

    try {
      const newArticle = await resolveUrls({
        path,
        taxonomyVersion: "default",
      });
      const splittedUri = newArticle.contentUri.split(":");
      const articleId = splittedUri[splittedUri.length - 1];
      navigate(routes.editArticle(parseInt(articleId), "standard"));
    } catch {
      navigate(routes.notFound);
    }
  };

  const handleSubmit = (evt: FormEvent) => {
    evt.preventDefault();
    const isNDLAUrl = isNDLAFrontendUrl(query);
    const isNodeId = query.length > 2 && /#\d+/g.test(query) && !Number.isNaN(parseFloat(query.substring(1)));

    const isTaxonomyId = query.length > 2 && /#urn:(resource|topic)[:\da-fA-F-]+/g.test(query);

    if (isNDLAUrl) {
      handleUrlPaste(query);
    } else if (isNodeId) {
      handleNodeId(parseInt(query.substring(1)));
    } else if (isTaxonomyId) {
      handleTaxonomyId(query.substring(1));
    } else {
      onSubmit();
    }
  };

  return (
    <StyledForm onSubmit={handleSubmit} {...props} ref={ref}>
      {children}
    </StyledForm>
  );
});
