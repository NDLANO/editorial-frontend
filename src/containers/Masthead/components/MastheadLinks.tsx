/**
 * Copyright (c) 2024-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ComponentPropsWithoutRef } from "react";
import { SafeLink } from "@ndla/safelink";
import { styled } from "@ndla/styled-system/jsx";
import { JsxStyleProps } from "@ndla/styled-system/types";

const LinksContainer = styled("div", {
  base: {
    display: "flex",
    gap: "xsmall",
  },
});

export const MastheadLinks = (props: ComponentPropsWithoutRef<"div"> & JsxStyleProps) => {
  return (
    <LinksContainer {...props}>
      <SafeLink target="_blank" to="https://edndla.zendesk.com/hc/no">
        Zendesk
      </SafeLink>
      <SafeLink target="_blank" to="https://kvalitet.ndla.no/">
        Kvalitaisen
      </SafeLink>
    </LinksContainer>
  );
};
