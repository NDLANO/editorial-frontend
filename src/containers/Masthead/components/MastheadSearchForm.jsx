/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Button from '@ndla/button';
import { colors } from '@ndla/core';
import { Search } from '@ndla/icons/common';
import { injectT } from '@ndla/i18n';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { css } from 'react-emotion';
import { toEditArticle, to404 } from '../../../util/routeHelpers';

import { fetchTopicArticle } from '../../../modules/taxonomy';

import { fetchDraft, fetchNewArticleId } from '../../../modules/draft/draftApi';
import { getLocale } from '../../../modules/locale/locale';
import { editorialMastheadClasses } from '../MastheadContainer';

export class MastheadSearchForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      query: props.query,
    };
    this.handleQueryChange = this.handleQueryChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleUrlPaste = this.handleUrlPaste.bind(this);
    this.handleNodeId = this.handleNodeId.bind(this);
  }

  handleQueryChange(evt) {
    this.setState({ query: evt.target.value });
  }

  async handleNodeId(nodeId) {
    const { history, locale } = this.props;
    try {
      const newArticle = await fetchNewArticleId(nodeId);
      const article = await fetchDraft(newArticle.id);
      const [supportedLanguage] = article.supportedLanguages.filter(
        item => item === locale,
      );

      history.push(
        toEditArticle(
          newArticle.id,
          'standard',
          supportedLanguage || article.supportedLanguages[0],
        ),
      );
    } catch (error) {
      history.push(to404());
    }
  }

  handleUrlPaste(ndlaUrl) {
    const { history, locale } = this.props;

    // Removes search queries before split
    const splittedNdlaUrl = ndlaUrl.split(/\?/)[0].split('/');

    const urlId = splittedNdlaUrl[splittedNdlaUrl.length - 1];
    if (!urlId.includes('urn:topic') && Number.isNaN(parseFloat(urlId))) return;

    this.setState({ query: '' });
    if (urlId.includes('urn:topic')) {
      fetchTopicArticle(urlId, locale).then(topicArticle => {
        const arr = topicArticle.contentUri.split(':');
        const id = arr[arr.length - 1];
        history.push(toEditArticle(id, 'topic-article', locale));
      });
    } else if (
      splittedNdlaUrl.includes('ndla.no') &&
      splittedNdlaUrl.includes('node')
    ) {
      this.handleNodeId(urlId);
    } else {
      history.push(toEditArticle(urlId, 'standard', locale));
    }
  }

  handleSubmit(evt) {
    evt.preventDefault();
    const { query } = this.state;
    const isNDLAUrl = /^https:\/(.*).ndla.no\/(article|subjects|nb|nn|en)\/(node|\d*)(\/|\d*)/.test(
      query,
    );
    const isNodeId =
      query.length > 2 &&
      /#\d+/g.test(query) &&
      !Number.isNaN(parseFloat(query.substring(1)));

    if (isNDLAUrl) {
      this.handleUrlPaste(query);
    } else if (isNodeId) {
      this.handleNodeId(query.substring(1));
    } else {
      this.props.onSearchQuerySubmit(query);
    }
  }

  render() {
    const { searching, t } = this.props;

    return (
      <form onSubmit={this.handleSubmit} {...editorialMastheadClasses('form')}>
        <input
          type="text"
          {...editorialMastheadClasses('form-query')}
          onChange={this.handleQueryChange}
          value={this.state.query}
          placeholder={t('searchForm.placeholder')}
        />
        <Button
          submit
          stripped
          css={css`
            color: ${colors.brand.grey};
          `}
          loading={searching}>
          <Search className="c-icon--medium" />
        </Button>
      </form>
    );
  }
}

MastheadSearchForm.propTypes = {
  locale: PropTypes.string.isRequired,
  query: PropTypes.string,
  searching: PropTypes.bool.isRequired,
  onSearchQuerySubmit: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
};

MastheadSearchForm.defaultProps = {
  query: '',
};

const mapStateToProps = state => ({
  locale: getLocale(state),
});

export default withRouter(
  connect(mapStateToProps)(injectT(MastheadSearchForm)),
);
