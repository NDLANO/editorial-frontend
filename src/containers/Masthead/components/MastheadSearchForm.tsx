/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { FormEvent, useState } from 'react';
import { ButtonV2 } from '@ndla/button';
import { useNavigate } from 'react-router-dom';
import { colors, misc, spacing, fonts } from '@ndla/core';
import { Search } from '@ndla/icons/common';
import { useTranslation } from 'react-i18next';
import styled from '@emotion/styled';
import { isValidLocale } from '../../../i18n';
import { toEditArticle, to404 } from '../../../util/routeHelpers';
import { isNDLAFrontendUrl } from '../../../util/htmlHelpers';
import { fetchResource, fetchTopic } from '../../../modules/taxonomy';
import { fetchNewArticleId } from '../../../modules/draft/draftApi';
import { resolveUrls } from '../../../modules/taxonomy/taxonomyApi';
import { useTaxonomyVersion } from '../../StructureVersion/TaxonomyVersionProvider';

const StyledForm = styled.form<{ color: string }>`
  display: flex;
  flex-grow: 1;
  background: ${(props) => props.color};
  border-radius: ${misc.borderRadius};
  border: 1px solid transparent;
  ${fonts.sizes(16, 1.2)};
  padding: ${spacing.xsmall};
  height: 48px;
  align-items: center;
  transition: all 100ms ease-in-out;

  &:focus-within {
    background: ${colors.white};
  }

  input {
    padding: ${spacing.xsmall};
    background: transparent;
    border: 0;
    outline: none;
    color: ${colors.brand.primary};
    transition: all 300ms ease-in-out;

    &:not(:focus-within) {
      ::placeholder {
        color: ${colors.brand.primary};
        font-weight: ${fonts.weight.semibold};
        opacity: 1;
      }
    }
  }

  & > button {
    color: ${colors.brand.grey};
  }

  .c-icon {
    margin: 0 ${spacing.small};
  }

  &:focus-within {
    border: 1px solid ${colors.brand.tertiary};

    .c-icon {
      color: ${colors.brand.primary};
    }
  }

  &:hover,
  &:focus {
    &:not(:focus-within) {
      cursor: pointer;
      border: 1px solid ${colors.brand.tertiary};
    }
  }
`;

const StyledSearch = styled(Search)`
  margin: 0 ${spacing.small};
  width: 24px;
  height: 24px;
`;

interface Props {
  query?: string;
  onSearchQuerySubmit: (query: string) => void;
  color: string;
}

export const MastheadSearchForm = ({
  query: initQuery = '',
  onSearchQuerySubmit,
  color,
}: Props) => {
  const [query, setQuery] = useState(initQuery);
  const { t, i18n } = useTranslation();
  const { taxonomyVersion } = useTaxonomyVersion();
  const navigate = useNavigate();

  const handleQueryChange = (evt: FormEvent<HTMLInputElement>) => setQuery(evt.currentTarget.value);

  const handleNodeId = async (nodeId: number) => {
    try {
      const newArticle = await fetchNewArticleId(nodeId);
      navigate(toEditArticle(newArticle.id, 'standard'));
    } catch (error) {
      navigate(to404());
    }
  };

  const handleTaxonomyId = async (taxId: string) => {
    const taxonomyFunction = taxId.includes('urn:resource:') ? fetchResource : fetchTopic;
    try {
      const taxElement = await taxonomyFunction({ id: taxId, taxonomyVersion });
      const arr = taxElement.contentUri?.split(':');
      if (arr) {
        const id = arr[arr.length - 1];
        navigate(toEditArticle(parseInt(id), 'standard'));
      }
    } catch (error) {
      navigate(to404());
    }
  };

  const handleUrlPaste = (frontendUrl: string) => {
    // Removes search queries before split
    const ndlaUrl = frontendUrl.split(/\?/)[0];
    // Strip / from end if topic
    const cleanUrl = ndlaUrl.endsWith('/')
      ? ndlaUrl.replace('/subjects', '').slice(0, -1)
      : ndlaUrl.replace('/subjects', '');
    const splittedNdlaUrl = cleanUrl.split('/');

    const urlId = splittedNdlaUrl[splittedNdlaUrl.length - 1];

    if (
      !urlId.includes('urn:topic') &&
      Number.isNaN(parseFloat(urlId)) &&
      !splittedNdlaUrl.find((e) => e.match(/subject:*/)) === undefined
    ) {
      return;
    }
    setQuery('');
    if (urlId.includes('urn:topic')) {
      handleTopicUrl(urlId);
    } else if (splittedNdlaUrl.includes('node')) {
      handleNodeId(parseInt(urlId));
    } else if (splittedNdlaUrl.find((e) => e.match(/subject:*/))) {
      handleFrontendUrl(cleanUrl);
    } else {
      navigate(toEditArticle(parseInt(urlId), 'standard'));
    }
  };

  const handleTopicUrl = async (urlId: string) => {
    try {
      const topicArticle = await fetchTopic({
        id: urlId,
        language: i18n.language,
        taxonomyVersion,
      });
      const arr = topicArticle.contentUri.split(':');
      const id = arr[arr.length - 1];
      navigate(toEditArticle(parseInt(id), 'topic-article'));
    } catch {
      navigate(to404());
    }
  };

  const handleFrontendUrl = async (url: string) => {
    const { pathname } = new URL(url);
    const paths = pathname.split('/');
    const path = isValidLocale(paths[1]) ? paths.slice(2).join('/') : pathname;

    try {
      const newArticle = await resolveUrls({ path, taxonomyVersion: 'default' });
      const splittedUri = newArticle.contentUri.split(':');
      const articleId = splittedUri[splittedUri.length - 1];
      navigate(toEditArticle(parseInt(articleId), 'standard'));
    } catch {
      navigate(to404());
    }
  };

  const handleSubmit = (evt: FormEvent) => {
    evt.preventDefault();
    const isNDLAUrl = isNDLAFrontendUrl(query);
    const isNodeId =
      query.length > 2 && /#\d+/g.test(query) && !Number.isNaN(parseFloat(query.substring(1)));

    const isTaxonomyId = query.length > 2 && /#urn:(resource|topic)[:\da-fA-F-]+/g.test(query);

    if (isNDLAUrl) {
      handleUrlPaste(query);
    } else if (isNodeId) {
      handleNodeId(parseInt(query.substring(1)));
    } else if (isTaxonomyId) {
      handleTaxonomyId(query.substring(1));
    } else {
      onSearchQuerySubmit(query);
    }
  };
  const [focused, setFocused] = useState(false);

  return (
    <StyledForm onSubmit={handleSubmit} color={color}>
      <ButtonV2 type="submit" variant="stripped">
        <StyledSearch color={colors.brand.primary} />
      </ButtonV2>

      <input
        type="text"
        onChange={handleQueryChange}
        value={query}
        placeholder={focused ? t('searchForm.placeholder') : t('searchPage.searchButton')}
        onFocus={() => setFocused(true)}
        onBlur={() => setFocused(false)}
      />
    </StyledForm>
  );
};

export default MastheadSearchForm;
