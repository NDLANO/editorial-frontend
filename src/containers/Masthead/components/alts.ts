/**
 * Copyright (c) 2025-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

const alts = {
  nb: [
    "Norske Dovendyrs Laug og Allianse",
    "Nyttige Dagens Lunsj Anbefalinger",
    "Nasjonal Dugnad for Latter og Anarki",
    "Nysgjerrige Detektivers Løsnings Arkiv",
    "Nordmenns Daglige Lykke Aktivitet",
    "Nattens Dansende Lyskastere Allianse",
    "Nyttige Data for Lærings Aktiviteter",
    "Norges Dyreste Luksus Artikler",
    "Nydelige Dyr Lager Abstraksjoner",
    "Nasjonal Dugnad for Løft av Alpakkaer",
  ],
  nn: [
    "Norske Dovendyrs Laug og Allianse",
    "Nyttige Dagens Lunsj Anbefalinger",
    "Nasjonal Dugnad for Latter og Anarki",
    "Nysgjerrige Detektivers Løsnings Arkiv",
    "Nordmenns Daglige Lykke Aktivitet",
    "Nattens Dansende Lyskastere Allianse",
    "Nyttige Data for Lærings Aktiviteter",
    "Norges Dyreste Luksus Artikler",
    "Nydelige Dyr Lager Abstraksjoner",
    "Nasjonal Dugnad for Løft av Alpakkaer",
  ],
  en: [
    "Never Drink Lemonade Alone",
    "Nocturnal Dancing Llama Association",
    "Ninja Donuts Love Adventures",
    "New Dinosaur Language Academy",
    "Notorious Duck Lovers Anonymous",
    "National Day of Lazy Activities",
    "Nerds Discussing Lizard Anatomy",
    "Neverending Dreams of Lush Adventures",
    "Nomadic Dragons Learning Algebra",
    "Nostalgic Doodles and Laughs Assembly",
  ],
};
export default alts;
