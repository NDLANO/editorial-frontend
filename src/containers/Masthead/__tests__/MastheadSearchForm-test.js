/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import nock from 'nock';
import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import { MastheadSearchForm } from '../components/MastheadSearchForm';

const noop = () => {};

test('MastheadSearchForm redirects on ndla url paste with id at the end', () => {
  const historyMock = {
    push: sinon.spy(),
  };

  const component = renderer.create(
    <MastheadSearchForm
      show
      query=""
      searching={false}
      onSearchQuerySubmit={noop}
      t={() => ''}
      locale="nb"
      history={historyMock}
    />,
  );
  const tree = component.toJSON();
  const e = {
    preventDefault: () => {},
  };
  tree.props.onSubmit(e);
  expect(component.toJSON()).toMatchSnapshot();
  setTimeout(() => {
    expect(historyMock.push.calledOnce).toBe(true);
    expect(historyMock.push.calledWith('/learning-resource/3333/edit/nb')).toBe(
      true,
    );
  }, global.DEFAULT_TIMEOUT);
});

test('MastheadSearchForm redirects on ndla url paste with taxonomy id at the end', () => {
  const historyMock = {
    push: sinon.spy(),
  };

  nock('http://ndla-api')
    .get('/taxonomy/v1/topics/urn:topic:1:179373/?language=nb')
    .reply(200, { contentUri: 'urn:content:4232' });

  const component = renderer.create(
    <MastheadSearchForm
      show
      query="https://ndla-frontend.test.api.ndla.no/article/urn:subject:100/urn:topic:1:179373"
      searching={false}
      locale="nb"
      onSearchQuerySubmit={noop}
      t={() => ''}
      history={historyMock}
    />,
  );
  const tree = component.toJSON();
  const e = {
    preventDefault: () => {},
  };
  tree.props.onSubmit(e);
  expect(component.toJSON()).toMatchSnapshot();
  return new Promise(resolve => {
    setTimeout(() => {
      expect(historyMock.push.calledOnce).toBe(true);
      expect(historyMock.push.calledWith('/topic-article/4232/edit/nb')).toBe(
        true,
      );
      resolve();
    }, global.DEFAULT_TIMEOUT);
  });
});

test('MastheadSearchForm redirects on old ndla url paste with new id', () => {
  const historyMock = {
    push: sinon.spy(),
  };

  nock('http://ndla-api')
    .get('/draft-api/v1/drafts/external_id/4737')
    .reply(200, { id: '123' });

  const component = renderer.create(
    <MastheadSearchForm
      show
      query="https://ndla.no/nb/node/4737?fag=36"
      searching={false}
      locale="nb"
      onSearchQuerySubmit={noop}
      t={() => ''}
      history={historyMock}
    />,
  );
  const tree = component.toJSON();
  const e = {
    preventDefault: () => {},
  };
  tree.props.onSubmit(e);
  expect(component.toJSON()).toMatchSnapshot();
  return new Promise(resolve => {
    setTimeout(() => {
      expect(historyMock.push.calledOnce).toBe(true);
      expect(historyMock.push.getCall(0).args[0]).toBe(
        '/learning-resource/123/edit/nb',
      );
      resolve();
    }, global.DEFAULT_TIMEOUT);
  });
});

test('MastheadSearchForm invalid id at the end of the url', () => {
  const historyMock = {
    push: sinon.spy(),
  };

  const component = renderer.create(
    <MastheadSearchForm
      show
      query="https://ndla-frontend.test.api.ndla.no/article/urn:subject:100/urn:topic:1:179373/urn:resource:1:16838"
      searching={false}
      onSearchQuerySubmit={noop}
      t={() => ''}
      locale="nb"
      history={historyMock}
    />,
  );
  const tree = component.toJSON();
  const e = {
    preventDefault: () => {},
  };
  tree.props.onSubmit(e);
  expect(component.toJSON()).toMatchSnapshot();
  setTimeout(() => {
    expect(historyMock.push.calledOnce).toBe(false);
  }, global.DEFAULT_TIMEOUT);
});

test('MastheadSearchForm redirects on ndla node id pasted', () => {
  const historyMock = {
    push: sinon.spy(),
  };
  nock('http://ndla-api')
    .get('/draft-api/v1/drafts/external_id/4737')
    .reply(200, { id: '123' });

  const component = renderer.create(
    <MastheadSearchForm
      show
      query="#4737"
      searching={false}
      locale="nb"
      onSearchQuerySubmit={noop}
      t={() => ''}
      history={historyMock}
    />,
  );
  const tree = component.toJSON();
  const e = {
    preventDefault: () => {},
  };
  tree.props.onSubmit(e);
  expect(component.toJSON()).toMatchSnapshot();
  return new Promise(resolve => {
    setTimeout(() => {
      expect(historyMock.push.calledOnce).toBe(true);
      expect(
        historyMock.push.calledWith('/learning-resource/123/edit/nb'),
      ).toBe(true);
      resolve();
    }, global.DEFAULT_TIMEOUT);
  });
});
