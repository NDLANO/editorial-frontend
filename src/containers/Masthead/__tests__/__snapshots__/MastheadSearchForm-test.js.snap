// Jest Snapshot v1, https://goo.gl/fbAQLP


exports[`MastheadSearchForm invalid id at the end of the url 1`] = `
<form
  className="css-1xnpzgm"
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder=""
    type="text"
    value="https://ndla-frontend.test.api.ndla.no/article/urn:subject:100/urn:topic:1:179373/urn:resource:1:16838"
  />
  <button
    className="css-e0v71u emotion-0"
    disabled={false}
    type="submit"
  >
    <svg
      fill="currentColor"
      preserveAspectRatio="xMidYMid meet"
      height="1em"
      width="1em"
      class="c-icon c-icon--medium"
      viewBox="0 0 48 48"
      data-license="Apache License 2.0"
      data-source="Material Design"
      style="vertical-align: middle;"
    >
      <title>Search</title>
      <g>
        <path
          d="M31 28h-1.59l-.55-.55C30.82 25.18 32 22.23 32 19c0-7.18-5.82-13-13-13S6 11.82 6 19s5.82 13 13 13c3.23 0 6.18-1.18 8.45-3.13l.55.55V31l10 9.98L40.98 38 31 28zm-12 0c-4.97 0-9-4.03-9-9s4.03-9 9-9 9 4.03 9 9-4.03 9-9 9z"
        />
      </g>
    </svg>
  </button>
</form>
`;

exports[`MastheadSearchForm redirects on ndla node id pasted 1`] = `
<form
  className="css-1xnpzgm"
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder=""
    type="text"
    value="#4737"
  />
  <button
    className="css-e0v71u emotion-0"
    disabled={false}
    type="submit"
  >
    <svg
      className="c-icon c-icon--medium"
      data-license="Apache License 2.0"
      data-source="Material Design"
      fill="currentColor"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "color": undefined,
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 48 48"
      width="1em"
    >
      <title>
        Search
      </title>
      <g>
        <path
          d="M31 28h-1.59l-.55-.55C30.82 25.18 32 22.23 32 19c0-7.18-5.82-13-13-13S6 11.82 6 19s5.82 13 13 13c3.23 0 6.18-1.18 8.45-3.13l.55.55V31l10 9.98L40.98 38 31 28zm-12 0c-4.97 0-9-4.03-9-9s4.03-9 9-9 9 4.03 9 9-4.03 9-9 9z"
        />
      </g>
    </svg>
  </button>
</form>
`;

exports[`MastheadSearchForm redirects on ndla url paste with id at the end 1`] = `
<form
  className="css-1xnpzgm"
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder=""
    type="text"
    value=""
  />
  <button
    className="css-e0v71u emotion-0"
    disabled={false}
    type="submit"
  >
    <svg
      className="c-icon c-icon--medium"
      data-license="Apache License 2.0"
      data-source="Material Design"
      fill="currentColor"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "color": undefined,
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 48 48"
      width="1em"
    >
      <title>
        Search
      </title>
      <g>
        <path
          d="M31 28h-1.59l-.55-.55C30.82 25.18 32 22.23 32 19c0-7.18-5.82-13-13-13S6 11.82 6 19s5.82 13 13 13c3.23 0 6.18-1.18 8.45-3.13l.55.55V31l10 9.98L40.98 38 31 28zm-12 0c-4.97 0-9-4.03-9-9s4.03-9 9-9 9 4.03 9 9-4.03 9-9 9z"
        />
      </g>
    </svg>
  </button>
</form>
`;

exports[`MastheadSearchForm redirects on ndla url paste with taxonomy id at the end 1`] = `
<form
  className="css-1xnpzgm"
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder=""
    type="text"
    value=""
  />
  <button
    className="css-e0v71u emotion-0"
    disabled={false}
    type="submit"
  >
    <svg
      className="c-icon c-icon--medium"
      data-license="Apache License 2.0"
      data-source="Material Design"
      fill="currentColor"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "color": undefined,
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 48 48"
      width="1em"
    >
      <title>
        Search
      </title>
      <g>
        <path
          d="M31 28h-1.59l-.55-.55C30.82 25.18 32 22.23 32 19c0-7.18-5.82-13-13-13S6 11.82 6 19s5.82 13 13 13c3.23 0 6.18-1.18 8.45-3.13l.55.55V31l10 9.98L40.98 38 31 28zm-12 0c-4.97 0-9-4.03-9-9s4.03-9 9-9 9 4.03 9 9-4.03 9-9 9z"
        />
      </g>
    </svg>
  </button>
</form>
`;

exports[`MastheadSearchForm redirects on old ndla url paste with new id 1`] = `
<form
  className="css-1xnpzgm"
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder=""
    type="text"
    value=""
  />
  <button
    className="css-e0v71u emotion-0"
    disabled={false}
    type="submit"
  >
    <svg
      className="c-icon c-icon--medium"
      data-license="Apache License 2.0"
      data-source="Material Design"
      fill="currentColor"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "color": undefined,
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 48 48"
      width="1em"
    >
      <title>
        Search
      </title>
      <g>
        <path
          d="M31 28h-1.59l-.55-.55C30.82 25.18 32 22.23 32 19c0-7.18-5.82-13-13-13S6 11.82 6 19s5.82 13 13 13c3.23 0 6.18-1.18 8.45-3.13l.55.55V31l10 9.98L40.98 38 31 28zm-12 0c-4.97 0-9-4.03-9-9s4.03-9 9-9 9 4.03 9 9-4.03 9-9 9z"
        />
      </g>
    </svg>
  </button>
</form>
`;
